diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,320 @@
+import abc
+from collections.abc import Callable
+from typing import Any, Generic, TypeVar
+
+import equinox as eqx
+import equinox.internal as eqxi
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+from equinox import AbstractVar
+from jax import vmap
+from jax.tree_util import PyTreeDef
+from jaxtyping import Array, ArrayLike, PyTree
+
+from .._custom_types import (
+    AbstractBrownianIncrement,
+    BoolScalarLike,
+    DenseInfo,
+    RealScalarLike,
+)
+from .._local_interpolation import LocalLinearInterpolation
+from .._solution import RESULTS
+from .._term import (
+    AbstractTerm,
+    LangevinDiffusionTerm,
+    LangevinDriftTerm,
+    LangevinTuple,
+    LangevinX,
+    MultiTerm,
+    WrapTerm,
+)
+from .base import AbstractItoSolver, AbstractStratonovichSolver
+
+
+_ErrorEstimate = TypeVar("_ErrorEstimate", None, LangevinTuple)
+_LangevinArgs = tuple[LangevinX, LangevinX, Callable[[LangevinX], LangevinX]]
+
+
+def get_args_from_terms(
+    terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
+) -> _LangevinArgs:
+    drift, diffusion = terms.terms
+    if isinstance(drift, WrapTerm):
+        assert isinstance(diffusion, WrapTerm)
+        drift = drift.term
+        diffusion = diffusion.term
+
+    assert isinstance(drift, LangevinDriftTerm)
+    assert isinstance(diffusion, LangevinDiffusionTerm)
+    gamma = drift.gamma
+    u = drift.u
+    f = drift.grad_f
+    return gamma, u, f
+
+
+class AbstractCoeffs(eqx.Module):
+    dtype: AbstractVar[jnp.dtype]
+
+
+_Coeffs = TypeVar("_Coeffs", bound=AbstractCoeffs)
+
+
+class SolverState(eqx.Module, Generic[_Coeffs]):
+    h: RealScalarLike
+    taylor_coeffs: PyTree[_Coeffs, "LangevinX"]
+    coeffs: _Coeffs
+    rho: LangevinX
+    prev_f: LangevinX
+
+
+# CONCERNING COEFFICIENTS:
+# The coefficients used in a step of this SRK depend on
+# the time increment h, and the parameter gamma.
+# Assuming the modelled SDE stays the same (i.e. gamma is fixed),
+# then these coefficients must be recomputed each time h changes.
+# Furthermore, for very small h, directly computing the coefficients
+# via the function below can cause large floating point errors.
+# Hence, we pre-compute the Taylor expansion of the SRK coefficients
+# around h=0. Then we can compute the SRK coefficients either via
+# the Taylor expansion, or via direct computation.
+# In short the Taylor coefficients give a Taylor expansion with which
+# one can compute the SRK coefficients more precisely for a small h.
+
+
+class AbstractLangevinSRK(
```

Comment:
The solvers of this type are all due to James, right? If so then I nominate that we call these `AbstractFosterLangevinSRK` instead :)

We tend to name solvers after the person who came up with them rather than the problem they solve! :)

Reply:
Yep, sound good :)))